[adorne] Adorne Binding initial contribution

theiding opened the pull request:
[adorne] Initial contribution

Closes #6011

Hilbrand commented:
Remove the generated javadoc. Generated files should in general not be committed.

theiding commented:
Sorry about the Javadocs. Removed now.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/discovery/AdorneDiscoveryService.java:
```diff
@@ -0,0 +1,112 @@
+            return adorneHubController.getZones();
+        }).thenApply(zoneIds -> {
+            for (int zoneId : zoneIds) {
+                adorneHubController.getState(zoneId).thenApply(state -> {
```
Same here. Maybe also use `zoneIds.forEach(...)`

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneDimmerHandler.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@NonNullByDefault
+public class AdorneDimmerHandler extends AdorneSwitchHandler {
+
```
remove empty line

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHandlerFactory.java:
```diff
@@ -0,0 +1,79 @@
+            logger.debug("Creating an AdorneSwitchHandler for thing '{}'", thing.getUID());
+
+            return new AdorneSwitchHandler(thing);
+        }
```
```suggestion
        } else if (...) {
```

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHandlerFactory.java:
```diff
@@ -0,0 +1,79 @@
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (thingTypeUID.equals(THING_TYPE_SWITCH)) {
+            logger.debug("Creating an AdorneSwitchHandler for thing '{}'", thing.getUID());
```
I think you can remove these log-statements

theiding replied:
I found these very helpful for troubleshooting as the framework doesn't log anything in this regard. I believe these log statements are useful given that they are at debug level.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHandlerFactory.java:
```diff
@@ -0,0 +1,79 @@
+@NonNullByDefault
+@Component(configurationPid = "binding.adorne", service = ThingHandlerFactory.class)
+public class AdorneHandlerFactory extends BaseThingHandlerFactory {
+
```
remove empty line

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHubHandler.java:
```diff
@@ -0,0 +1,142 @@
+
+    public AdorneHubHandler(Bridge bridge) {
+        super(bridge);
+        this.bridge = bridge;
```
No need to cache the bridge

theiding replied:
Good point. Done. Applied the forEach here as well.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHubHandler.java:
```diff
@@ -0,0 +1,142 @@
+        AdorneHubConfiguration config = getConfigAs(AdorneHubConfiguration.class);
+        logger.debug("Configuration host:{} port:{}", config.host, config.port);
+
+        AdorneHubController adorneHubController = this.adorneHubController = new AdorneHubController(config.host,
```
Please make this two statements.

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneSwitchHandler.java:
```diff
@@ -0,0 +1,184 @@
+        logger.trace("handleCommand (channelUID:{} command:{}", channelUID, command);
+        try {
+            if (channelUID.getId().equals(CHANNEL_POWER)) {
+                if (command.equals(OnOffType.ON) || command.equals(OnOffType.OFF) || command instanceof RefreshType) {
```
```suggestion
                if (command.equals(OnOffType.ON) || command.equals(OnOffType.OFF)) {
```

and then later `} else (if command instanceof RefreshType`

theiding replied:
Yikes, how did I come up with that one? Good catch. Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneSwitchHandler.java:
```diff
@@ -0,0 +1,184 @@
+    public void refreshOnOff() {
+        // Asynchronously get our onOff state from the hub controller and update our state accordingly
+        AdorneHubController adorneHubController = getAdorneHubController();
+        adorneHubController.getState(zoneId).thenApply(state -> {
```
```suggestion
        adorneHubController.getState(zoneId).thenRun(state -> {
```

theiding replied:
See above

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+    private @Nullable AdorneHubChangeNotify changeListener;
+    private Gson gson;
+
+    private static class ZoneIDFutureRecord {
```
Why is this static?

theiding replied:
Since ZoneIDFutureRecord is just a DTO to pass state through the futures it doesn't need to be an inner class that I then have to instantiate from an outer instance. Just keeps the usage a bit lighter weight.

J-N-K replied:
Correct.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+    public AdorneHubController(@Nullable String hubHost, @Nullable Integer hubPort,
+            ScheduledExecutorService scheduler) {
+        hubController = null;
+        this.hubHost = (hubHost == null) ? HUB_DEFAULT_HOST : hubHost;
```
Not needed if you provide the defaults in the config (see above)

theiding replied:
As mentioned above I feel it's important to centralize default definition/handling  in the AdorneHubController and not let it leak to a DTO.

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+        hubOut = null;
+        hubIn = null;
+        hubControllerConnected = new CompletableFuture<>();
+        ;
```
remove

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+        commandId = 0;
+
+        changeListener = null;
+        gson = new Gson();
```
Do you need a new Gson-instance or could you share one from the handler factory and pass it though the handler's constructor to the controller?

theiding replied:
Yes, I could reuse the gson instance. But using Gson is an implementation detail of the hub controller that the factory doesn't need to know about. I could just lazy load the gson initialization into a static field in the controller constructor. That way I get re-use while containing logic in the controller.
That said, I did a quick perf test and the initialization takes less than 1ms on my RaspberryPi (since I am not using any custom JSON models overhead is minimal) . Also, savings would only apply to the edge case of running multiple hubs. Is it worth adding code complexity for this optimization?

J-N-K replied:
If it is unlikely that someone uses more than one hub, I'm ok with the way it is.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+                    processMsgSystemInfo(hubMsg);
+                }
+            }
+        } catch (Exception e) {
```
Please do not catch `Exception`, catch specific exceptions that can apply here. (also applies below)

theiding replied:
All specific exceptions are caught in the respective methods in the loop. This is truly a generic error handler - any exception coming through is fatal, disconnects the hub and is logged as Error. Isn't that the desired behavior over letting them bubble up further?

Hilbrand replied:
Then it would be `RuntimeException` instead of `Exception`

theiding replied:
Good point. Done.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+        try {
+            synchronized (this) {
+                if (hubSocket == null) {
+                    Socket hubSocket = this.hubSocket = new Socket(hubHost, hubPort);
```
Please  split this in two statements (also applies below)

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+            try {
+                logger.debug("Couldn't establish hub connection ({}). Sleep {} ms before re-trying.", e.getMessage(),
+                        hubReconnectSleep);
+                Thread.sleep(hubReconnectSleep);
```
Don't use `sleep`. Schedule the next execution with `scheduler.schedule(...)`

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,546 @@
+                    removeStateCommands.add(stateCommand);
+                }
+            }
+            for (AdorneHubController.ZoneIDFutureRecord stateCommand : removeStateCommands) {
```
```suggestion
stateCommands.removeAll(removeStateCommands);
```

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/resources/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,27 @@
+		<description>The Adorne Hub serves as the bridge to control Adorne switches, dimmer switches and outlets.</description>
+
+		<config-description>
+			<parameter name="host" type="text" required="false">
```
```suggestion
			<parameter name="host" type="text">
```

and add `<default>(value that I have forgotten, LCM1.local?)</default>`

same for port

theiding replied:
Similar to the argument above I would like to keep default handling in the hub controller. And as a user I prefer not to see any value for an optional parameter - if there is a value then I need to think whether it applies to me.

theiding replied:
So I moved the default handling to AdorneHubConfiguration as per your suggestion. But since host and port are optional parameters I feel the UI should reflect that accordingly. If the user doesn't provide a value then it should come from AdorneHubConfiguration, not the UI.

J-N-K replied:
Usually we provide the same value in the configuration and the XML.

theiding replied:
Ok - made the change for consistency

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/resources/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,27 @@
+			<parameter name="port" type="integer" required="false">
+				<label>Port</label>
+				<description>
+					Port 
```
Remove if you don't provide additional information

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,52 @@
+		<config-description>
+			<parameter name="zoneId" type="integer" required="true">
+				<label>Zone ID</label>
+				<description>
```
see above

theiding replied:
Done

J-N-K commented on bundles/pom.xml:
```diff
@@ -290,7 +291,7 @@
   </dependencies>
 
   <properties>
```
remove if not needed

theiding replied:
This part is not owned/related to my binding, is it?

J-N-K replied:
correct

theiding commented:
@J-N-K , thank you for the feedback! Process question: 
What is the expected format for me to submit the requested changes? One commit per suggestion or a single commit for all of them? 
How do you verify that I indeed addressed all your feedback - should I mention each of your requests in the matching commit comment?
Where you specified a concrete change I can accept those directly in the GitHub UI, right? All in one batch or one commit per requested change?
Sorry for all the questions - I couldn't find any documentation on this. If you do have a pointer to process/best practices documentation please let me know.

Hilbrand commented:
You can make it one commit. That is probably the most simple approach. You can simply state something like 'Addressed review comments' in your commit message, and don't forget to sign off your commit. If you push changes for review comments GitHub will show an outdated badge on the review comment if the source line the comment was about was changed. That is one way we can see it changed. You can also click the 'Resolve conversation' button, but I personally find that as a reviewer less convenient as I than have to open each comment to see what I wrote.
Better to make the changes locally instead of applying it on GitHub otherwise you would still need to pull the changes back to you locally to test them. So it's simpler to just make the changes locally.

theiding commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,96 @@
+
+The Adorne Binding integrates [Adorne Wi-Fi ready devices](https://www.legrand.us/adorne/products/wireless-whole-house-lighting-controls.aspx) (switches, dimmers, outlets) from [Legrand](legrand.com).
+
+Legrand attempted to provide a public API based on Samsung's ARTIK Cloud and the initial version of this binding was based on that API. However, Samsung shut down ARTIK Cloud shortly after the release and Legrand has not offered a public API replacement since. That leaves direct interaction with the Adorne Hub as the only control option. Consequently the OpenHAB server and the Adorne Hub must be  ...
```
Done

theiding commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,96 @@
+
+## Discovery
+
+Auto-discovery is supported as long as the hub can be discovered using the default host and port. If the hub requires custom host and/or port configuration then setup through a .things file is needed. 
```
Applied

theiding replied:
Since the PaperUI discovery process doesn't allow user supplied parameters I can't discover via PaperUI if host and port are needed to discover the hub.

theiding commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,96 @@
+The hub offers two optional configuration parameters:
+|Parameter|Description|
+|-|-|
+|Host|The URL to reach the hub. The hub makes itself known through mDNS as `LCM1.local` and the host parameter defaults to this value. As long as the OpenHAB server and the hub are on the same broadcast domain for mDNS the host parameter doesn't need to be specified. |
```
Updated host as well as port in readme.

theiding commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,96 @@
+|Host|The URL to reach the hub. The hub makes itself known through mDNS as `LCM1.local` and the host parameter defaults to this value. As long as the OpenHAB server and the hub are on the same broadcast domain for mDNS the host parameter doesn't need to be specified. |
+|Port|The port the hub communicates on. By default the hub answers on port 2112 and the port parameter defaults to this value. As long as the hub configuration hasn't been changed the port parameter doesn't need to be specified.|
+
+Example if specific configuration values need to be provided:
```
Done. Merged the example with the existing example.

theiding commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,96 @@
+|power|Switch|ON, OFF|Turn device on and off|switch, dimmer|
+|brightness|Dimmer|1-100|Set device's brightness|dimmer|
+
+Note that the brightness channel is limited to values from 1 to 100. All other commands are ignored. That means in particular that a dimmer can't be turned off by sending 0 to the brightness channel. Also, if a dimmer is turned off (via the power channel) and the brightness is updated the dimmer will remain off. Once the dimmer is turned on it will turn on with the updated brightness setting. Con ...
```
It is 1-100 indeed. Please see the comments for the channel for the explanation. This behavior matches the semantics of the physical switch (as well as the Legrand app). 
No change made.

theiding commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,96 @@
+	}
+	Frame label="Adorne Dimmer using Slider" {
+		Switch item=LightBedroomSwitch1 label="Bedroom 1" mappings=["ON"="On", "OFF"="Off"] icon="light-on"
+		Slider item=LightBedroomDimmer1 label="Bedroom 1" icon="light-on" minValue=1 maxValue=100 step=1 // Requires OpenHAB 2.5
```
Done

theiding commented on bundles/org.openhab.binding.adorne/pom.xml:
```diff
@@ -0,0 +1,16 @@
+
+  <artifactId>org.openhab.binding.adorne</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: adorne Binding</name>
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/demo/items/demo.items:
```diff
@@ -0,0 +1,5 @@
+Switch LightBathroom {channel="adorne:switch:home:bathroom:power"}
```
Short of having automated tests I use these for manual testing. Since tests get checked in I felt these files should be checked in as well. 
No change made.

J-N-K replied:
If it's needed for tests, move to  `src/test/resources`

theiding commented on bundles/org.openhab.binding.adorne/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<features name="org.openhab.binding.adorne-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+    <repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${project.version}/xml/features</repository>
```
Done

theiding replied:
Actually, after I made this change I started getting the following build error. Do you know what that means?

[INFO] --- karaf-maven-plugin:4.2.7:verify (karaf-feature-verification) @ org.openhab.binding.adorne ---
[INFO] Using repositories: https://openhab.jfrog.io/openhab/libs-release@id=openhab-release,https://openhab.jfrog.io/openhab/libs-snapshot@id=openhab-snapshot@noreleases@snapshots,https://repo.maven.apache.org/maven2@id=central
[WARNING] Error resolving artifact org.openhab.core.features.karaf:org.openhab.core.features.karaf.openhab-core:jar:3.0.0-SNAPSHOT: [Could not find artifact org.openhab.core.features.karaf:org.openhab.core.features.karaf.openhab-core:jar:3.0.0-SNAPSHOT in openhab-snapshot (https://openhab.jfrog.io/openhab/libs-snapshot/)]
java.io.IOException: Error resolving artifact org.openhab.core.features.karaf:org.openhab.core.features.karaf.openhab-core:jar:3.0.0-SNAPSHOT: [Could not find artifact org.openhab.core.features.karaf:org.openhab.core.features.karaf.openhab-core:jar:3.0.0-SNAPSHOT in openhab-snapshot (https://openhab.jfrog.io/openhab/libs-snapshot/)]

J-N-K replied:
The problem is the wrong base branch. Your PR is against master (which is on 3.0.0). You need to rebase on `2.5.x`. If unsure how to do that, please ask, otherwise chances are good you mess up the whole branch. I can assist if needed.

theiding replied:
Worked fine as you said after I rebased. Done.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/configuration/AdorneHubConfiguration.java:
```diff
@@ -0,0 +1,27 @@
+ */
+@NonNullByDefault
+public class AdorneHubConfiguration {
+    public @Nullable String host;
```
AdorneHubConfiguration is just a DTO that has no knowledge of what the default value is.  Default management is solely handled in AdorneHubController.

J-N-K replied:
I disagree here. Setting the default in the thing configuration class is a common pattern in OH and should be used if applicable.

theiding replied:
Np, let's agree to disagree. If this is already a common pattern then I'll follow it. Implemented your guidance.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/configuration/AdorneHubConfiguration.java:
```diff
@@ -0,0 +1,27 @@
+@NonNullByDefault
+public class AdorneHubConfiguration {
+    public @Nullable String host;
+    public @Nullable Integer port;
```
Same response as above.

theiding replied:
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/discovery/AdorneDiscoveryService.java:
```diff
@@ -0,0 +1,112 @@
+            // We have fully discovered the hub
+            thingDiscovered(DiscoveryResultBuilder.create(bridgeUID[0]).withLabel(DISCOVERY_HUB_LABEL).build());
+            return adorneHubController.getZones();
+        }).thenApply(zoneIds -> {
```
Not sure I understand. The previous stage does return the zoneIds and the thenApply works on the zoneIds.

J-N-K replied:
Sorry. My fault. `.thenAccept` would be correct, as you use the result from the previous stage and don't produce a new output.

theiding replied:
Got it. Done.

theiding commented:
I have now addressed the review feedback, but there are a couple unresolved conversations left. Should I go ahead and commit what I have at this point or do you want to finalize unresolved conversations first?

J-N-K commented:
Yes please, commit what you already have. That makes it easier to track the changes.

theiding commented:
Thanks, just committed. I set all of our conversations that I addressed to resolved and left the ones with open questions as unresolved.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/feature/feature.xml:
```diff
@@ -1,9 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<features name="org.openhab.binding.adorne-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
-    <repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${project.version}/xml/features</repository>
+<features name="org.openhab.binding.adorne-${project.version}"
```
This is due to an old version of the spotless configuration. Please have one line per tag, regardless of the line-length. Otherwise the feature-aggregation will fail.

theiding replied:
Done

J-N-K replied:
Still on two lines

```suggestion
<features name="org.openhab.binding.adorne-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
```

theiding replied:
sorry, didn't push yet - unresolving comment.

theiding replied:
Pushed

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHubHandler.java:
```diff
@@ -105,16 +102,15 @@ public void dispose() {
+        getThing().getThings().forEach(thing -> {
+            AdorneSwitchHandler thingHandler = (AdorneSwitchHandler) thing.getHandler();
             if (thingHandler != null && thingHandler.getZoneId() == zoneId) {
                 thingHandler.updateState(CHANNEL_POWER, onOff ? OnOffType.ON : OnOffType.OFF);
```
```suggestion
                thingHandler.updateState(CHANNEL_POWER, OnOffType.from(onOff));
```

theiding replied:
I originally had the for loop, but changed it to forEach as per your suggestion above for zoneIDs. Switched this one back. 
Nice shortcut for the onOff - done.

J-N-K replied:
Did you forget to push these changes?

theiding replied:
sorry, didn't push yet - unresolving comment.

theiding replied:
Pushed

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneSwitchHandler.java:
```diff
@@ -148,20 +146,10 @@ public int getZoneId() {
         AdorneHubController adorneHubController = getAdorneHubController();
-        adorneHubController.getState(zoneId).thenApply(state -> {
+        adorneHubController.getState(zoneId).thenAccept(state -> {
             OnOffType onOffState = state.onOff ? OnOffType.ON : OnOffType.OFF;
```
```suggestion
            OnOffType onOffState = OnOffType.from(state.onOff);
```

theiding replied:
Done

theiding replied:
sorry, didn't push yet - unresolving conversation.

theiding replied:
Pushed

theiding commented:
Thanks for the additional feedback - I made the changes. Re. the rebase: I issued git fetch upstream and git rebase upstream/master. I got two conflicts during rebase - for bom/openhab-addons/pom.xml and bundles/pom.xml. If I understand correctly listing the binding details in the pom is no longer needed. Since I no longer needed my changes I resolved the conflicts by issuing git rm bom/openhab-addons/pom.xml and git rm bundles/pom.xml. The rebase then completed successfully. However, I expected that the rm just removed the 2 files from the commits and that I would have the latest version from upstream/master in my local branch (so that I can then locally test and push to origin afterwards), however, the two files got deleted entirely from my local branch. How do I sync them back? 
Also, I am assuming that the need for the binding entries in pom files got completely eliminated and not just moved to a different file, is that correct? If so I will update the binding development docs accordingly.

J-N-K commented:
You need to rebase on `2.5.x`,

So

```
git rebase upstream/2.5.x
```

Probably you have to resolve some merge conflicts then and force push (maybe `--force-with-lease`). I'll change the base branch of the  PR now, this will create a very bad PR, after your force push they should vanish.

J-N-K commented on bundles/org.openhab.binding.adorne/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>2.5.0-SNAPSHOT</version>
```
```suggestion
    <version>2.5.2-SNAPSHOT</version>
```

theiding replied:
Done

theiding replied:
Pushed

theiding commented:
@J-N-K , you asked me to rebase against 2.5.x? Does that mean my work will be added to 2.5.x? I thought 2.5.x would not receive any new features and that my work would be released under 3.0. 
I did rebase with upstream/master - when you say things are looking good now what are you referring to? 
The base branch that you switched for my PR is that something I should have done? When you say it will create a very bad PR what does that mean? 
Side note: I marked some of your comments as completed, but didn't push the changes due to the rebase issue. Shouldn't have marked the comments as complete until the actual push. Sorry about that - will fix it. 
Thank you for all your patience!

J-N-K commented:
Maintainers will make sure that all bindings in 2.5.x make it into 3.0 once we have a running 3.0 core, so 2.5.x is the correct base-branch. This will solve your local build problems (with the ohc.version) and also make your binding available for 2.5.x once it is merged. The "no new features" is true for the core, not the addons.

theiding commented:
@J-N-K , that certainly makes it much easier for me that I only need to rebase on 2.5.x. Did that and there were no conflicts at all. I just pushed changes for all your remaining feedback and resolved all conversations. 
Please note that the spotless:check insisted on a format change for src\main\history\dependencies.xml similar for what we discussed for features.xml where it put the features text on two lines. Is that ok?

J-N-K commented:
That file is not committed anyway, so no problem with the formatting. I think you need to change to 2.5.3-SNAPSHOT in the pom to satisfy Travis.

theiding commented:
Thanks for the pointer on Travis - couldn't figure out what happened. My first rebase still pulled in 2.5.2 so I had to do another rebase to get to 2.5.3 and move my binding version forward then as well. Now Travis passed.

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+            zones.add(zoneIdMap.get(HUB_TOKEN_ZID).intValue());
+        }
+        synchronized (this) {
+            for (CompletableFuture<List<Integer>> zoneCommand : zoneCommands) {
```
zoneCommands.forEach(zoneCommand -> zoneCommand.complete(zones));

Please also check other occurences of for-loops where you only iterate over the list to execute a command on each member.

theiding replied:
Fixed all of them

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+    private int hubPort;
+    private @Nullable Socket hubSocket;
+    private @Nullable PrintStream hubOut;
+    private @Nullable Scanner hubIn;
```
Don't use `Scanner` everything you read through it can't be garbage collected. I found that out the hard way.
I fiddled around and found that JsonStreamParser is actually a suitable replacement.

theiding replied:
I had no idea. Made the switch. Had to swap all the downtream Json handling/parsing accordingly - is also a bit cleaner as it's no longer just Map/String based.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/discovery/AdorneDiscoveryService.java:
```diff
@@ -0,0 +1,112 @@
+        }).thenAccept(zoneIds -> {
+            zoneIds.forEach(zoneId -> {
+                adorneHubController.getState(zoneId).thenAccept(state -> {
+                    // Strip zone ID's name to become a valid ThingUID
```
Where do I find UIDUtils?

J-N-K replied:
org.eclipse.smarthome.core.util

https://github.com/openhab/openhab-core/blob/2.5.0/bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/util/UIDUtils.java

If you type UIDUtils you should get the correct suggestions for import (at least IntelliJ does, not sure about Eclipse)

theiding replied:
Easier indeed. Done.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+        }
+    }
+
+    public AdorneHubController(AdorneHubConfiguration config, ScheduledExecutorService scheduler) {
```
I was never concerned about performance as the impacted calls are inherently slow. I figured synchronizing on this was the simplest. But your suggestions did indeed make the code much more readable. In fact it enabled me to find a bug in the stopping logic that I couldn't see before. Thank you for the pointers.

theiding commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,113 @@
+In other words power and brightness states are controlled independently.
+This matches how power and brightness are managed on the physical dimmer itself. 
+
+To avoid confusion for the user any UI must ensure that only values from 1 to 100 are passed to the brightness channel.
```
Any UI can enforce passing only the supported values (e.g. Slider item=LightBedroomDimmer1 label="Bedroom 1" icon="light-on" minValue=1 maxValue=100 step=1). It's just the reality that 0 has no meaning for this channel (just like -1 or 1000 don't have a meaning and must not be passed in) and a well written UI can't ignore that. Letting the user select something invalid and then swallowing it silently without an error or morphing it into something the user didn't pick would both be confusing to the user. I added additional clarification in the readme why a UI that allows 0 should not be used.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+        // Shut down
+        synchronized (this) {
+            // If there are still pending requests we need to cancel them
+            for (AdorneHubController.ZoneIDFutureRecord stateCommand : stateCommands) {
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+                if (hubSocket != null) {
+                    hubSocket.close();
+                }
+            } catch (Exception e) {
```
Done. Also removed unneeded try catch for hubIn, hubOut closing

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/discovery/AdorneDiscoveryService.java:
```diff
@@ -0,0 +1,112 @@
+
+        // Hack - we wrap the ThingUID in an array to make it appear effectively final to the compiler throughout the
+        // chain of futures. Passing it through the chain as context would bloat the code.
+        ThingUID bridgeUID[] = new ThingUID[1];
```
Done. I guess my C roots are leaking through...

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+ * @author Mark Theiding - Initial Contribution
+ */
+@NonNullByDefault
+public class AdorneHubController implements Runnable {
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+            } catch (Exception e) {
+                logger.warn("Closing hub controller output stream failed ({})", e.getMessage());
+            }
+            ;
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+            } catch (Exception e) {
+                logger.warn("Closing hub controller input stream failed ({})", e.getMessage());
+            }
+            ;
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+            macAddressCommands.add(macAddressCommand);
+        }
+        sendRestCmd(String.format(HUB_REST_REQUEST_MACADDRESS, getNextCommandId()));
+        return macAddressCommand;
```
Very good point. Fixed same for zoneIdCommand and stateCommands. Much clearer now that the same server response is used to answer duplicate commands. For stateCommands a HashMap also much better abstracts the zoneId to future mapping than then old Lists. Allowed me also to get rid of the inner ZoneIDFutureRecord class.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,550 @@
+                    // Re-start message loop after sleep time
+                    synchronized (this) {
+                        hubController = scheduler.schedule(this, sleep, TimeUnit.SECONDS);
+                    }
```
Fixed as part of the overall locking refactor

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/discovery/AdorneDiscoveryService.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    public AdorneDiscoveryService() {
+        // Passing false as last argument to super constructor turns off background discovery
+        super(new HashSet<>(Arrays.asList(new ThingTypeUID(BINDING_ID, "-"))), DISCOVERY_TIMEOUT_SECONDS, false);
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/discovery/AdorneDiscoveryService.java:
```diff
@@ -0,0 +1,112 @@
+        ThingUID bridgeUID[] = new ThingUID[1];
+
+        // Future enhancement: Need a timeout for each future execution to recover from bugs in the hub controller, but
+        // Java8 doesn't yet offer that
```
I war referring to that ideally the discovery service shouldn't trust the hub controller and protect itself from hub controller bugs.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHubHandler.java:
```diff
@@ -0,0 +1,137 @@
+     * The {@link AdorneHubHandler} is notified that the state of one of its physical devices has changed. The
+     * {@link AdorneHubHandler} then asks the appropriate thing handler to update the thing to match the new state.
+     *
+     */
```
They do, right? The hub is notified as the central gatekeeper and then passed the notification on to the appropriate handlers so they can deal with their respective things.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHubHandler.java:
```diff
@@ -0,0 +1,137 @@
+                    thingHandler.refresh();
+                }
+            });
+            updateStatus(ThingStatus.ONLINE);
```
I can't find anything about such a problem in the docs. It also appears to be working fine (turn light on, disconnect hub, turn light off, reconnect hub, UI updates light as off). I believe this is ok.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneSwitchHandler.java:
```diff
@@ -0,0 +1,172 @@
+        logger.trace("handleCommand (channelUID:{} command:{}", channelUID, command);
+        try {
+            if (channelUID.getId().equals(CHANNEL_POWER)) {
+                if (command.equals(OnOffType.ON) || command.equals(OnOffType.OFF)) {
```
Done

theiding commented:
Sorry for the delay. Finally made it through the feedback. I especially liked the result of the synchronization changes. Thank you for taking time to provide such detailed and constructive feedback!

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+                }
+            }
+        } catch (RuntimeException e) {
+            logger.error("Hub controller failed", e);
```
logging error should be reserved for issues catastrophic enough severely impact the operation of openHab. I don't think a controller error in a binding would qualify as such.
```suggestion
            logger.warn("Hub controller failed", e);
```

theiding replied:
Done

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+        logger.info("Starting hub controller");
+        hubController = scheduler.submit(() -> {
+            msgLoop();
+        });
```
```suggestion
        hubController = scheduler.submit(this::msgLoop);
```

theiding replied:
Done

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+        boolean commandsServed = stopWhenCommandsServed && stateCommandsIsEmpty && (zoneCommand == null)
+                && (macAddressCommand == null);
+
+        return (stop || timeoutExceeded || commandsServed);
```
```suggestion
        Future<?> hubController = this.hubController;
        return (hubController == null || hubController.isCancelled() || timeoutExceeded || commandsServed);
```

theiding replied:
Done as part of above stop refactor

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+                macAddressCommand.cancel(false);
+                this.macAddressCommand = null;
+            }
+        }
```
As an example of using an `AtomicReference<@Nullable CompletableFuture<String>>, this code could be changed to:
```suggestion
            CompletableFuture<String> macAddressCommand = this.macAddressCommandRef.getAndSet(null);
            if (macAddressCommand != null) {
                macAddressCommand.cancel(false);
            }
```
If you added this utility method:
```
private void cancel(boolean interrupt, @Nullable Future<?> future){
    if(future != null){
         future.cancel(interrupt);
    }
}
```
Then the suggestion above becomes even easier.
```suggestion
            cancel(false, this.macAddressCommandRef.getAndSet(null));
```

The cancel utility method can be used for cancelling other futures as well and saves a tiny bit of code. If used enough the refactoring becomes worth it.

theiding replied:
See my notes above why I didn't remove the macAddressCommandLock. For this section an AtomicReference works well, but I found it not to work well for the others.

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+    private @Nullable Socket hubSocket;
+    private @Nullable PrintStream hubOut;
+    private @Nullable InputStreamReader hubInReader;
+    private @Nullable JsonStreamParser hubIn;
```
If you put these fields into a new class and populated them as part of construction you could make them non-null and final thus doing away with all the null checks. In other bindings they typically call this class the xxxCommunicator or xxxConnection. WDYT?

theiding replied:
Good idea. Null checks still have to happen on the new connection object, but a connection class nicely groups the 4 attributes (including closing a connection). I definitely like the result.

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+            if (!stop) {
+                hubController = scheduler.schedule(() -> {
+                    msgLoop();
+                }, sleep, TimeUnit.SECONDS);
```
```suggestion
                hubController = scheduler.schedule(this::msgLoop, sleep, TimeUnit.SECONDS);
```

theiding replied:
Done

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+    private final Object macAddressCommandLock = new Object();
+    private final Object zoneCommandLock = new Object();
+
+    public AdorneHubController(AdorneHubConfiguration config, ScheduledExecutorService scheduler) {
```
Why don't you include `changeListener` as a construction parameter? That way you can make it non-null and final.

theiding replied:
It seems a bit heavy handed to require a listener, but it certainly makes the code simpler (also removed the setter). Drawback is that I had to implement a dummy listener for the discovery service, but not a big deal.

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+    private Map<Integer, CompletableFuture<AdorneDeviceState>> stateCommands;
+    private @Nullable CompletableFuture<List<Integer>> zoneCommand;
+    private @Nullable CompletableFuture<String> macAddressCommand;
+    private int commandId; // We assign increasing command ids to all rest commands to the hub for easier
```
Make this an AtomicInteger and just make calls to incrementAndGet() every time you want a new id. This approach is a much more thread safe.

theiding replied:
I hadn't considered parallel commands, but you are correct - I need to. However, I can't use incrementAndGet as that overflows into negative values. So I used updateAndGet. Looking at the result wouldn't a simple lock be much easier to read though?

cpmeister replied:
I think using an AtomicInteger displays clearer intentions as to how the field is used and accessed vs using basic synchronization. Also your use case happens to be exactly what AtomicInteger was designed for. I have no problem with you using updateAndGet to keep values positive. 👍 
But just to keep things in perspective, if you incremented the id once every second it would take over 63 years before it overflowed. 😮

theiding replied:
OpenHAB will be around then, right? :-)

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+     * @return a future for the list of zone IDs
+     */
+    public CompletableFuture<List<Integer>> getZones() {
+        sendRestCmd(String.format(HUB_REST_REQUEST_ZONES, getNextCommandId()));
```
Why are you already sending a rest command if a completable future for this command might already exist?

theiding replied:
See above

cpmeister commented on bundles/org.openhab.binding.adorne/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>2.5.3-SNAPSHOT</version>
```
A new release just went out so we need to change this again.
```suggestion
    <version>2.5.4-SNAPSHOT</version>
```

theiding replied:
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+     * @return a future for the {@link AdorneDeviceState}
+     */
+    public CompletableFuture<AdorneDeviceState> getState(int zoneId) {
+        sendRestCmd(String.format(HUB_REST_REQUEST_STATE, getNextCommandId(), zoneId));
```
There is no guaranteed processing of requests and they can get lost. I am sending the REST command each time for optimal resiliency. However, I only need to track one future for all requests of the same type.

cpmeister replied:
If requests can get lost why can't this class automatically retry if you don't get a response within a period of time?
It seems strange to push the burden of retrying onto the callers of this method, it isn't a scalable nor robust approach.

theiding replied:
I simply didn't want to add that level of complexity to the code. My thinking was: A lost message is very rare - I can accept a lost message, but not a permanently failed binding.

cpmeister replied:
Well if it working and such a change risks breaking it then keep it as it is.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+
+    // When we submit commmands to the hub we don't correlate commands and responses. We simply use the first available
+    // response that answers our question. For that we maintain lists of all pending commands.
+    private Map<Integer, CompletableFuture<AdorneDeviceState>> stateCommands;
```
Sure can be. Done.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+    private volatile @Nullable AdorneHubChangeNotify changeListener;
+
+    private final Object stopLock = new Object();
+    private final Object hubOutLock = new Object();
```
By protecting the hubOut with the lock I can guarantee that sendRestCmd doesn't print to a closed stream and always throws a well defined/meaningful exception if invoked before or after the hub is connected. Without the lock I would lose that. Doesn't that justify the explicit lock?

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+
+    private final Object stopLock = new Object();
+    private final Object hubOutLock = new Object();
+    private final Object stateCommandsLock = new Object();
```
Good point. Done.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+    private final Object stopLock = new Object();
+    private final Object hubOutLock = new Object();
+    private final Object stateCommandsLock = new Object();
+    private final Object macAddressCommandLock = new Object();
```
In getMACAddress I need to be able to create a new future and send the rest command if the current macAddressCommand is null, otherwise return current macAddressCommand. How do I do that with an AtomicReference? I got close with an updateAndGet, but it became very awkward. Isn't the simple lock I have more straight forward? Or is there a better way of doing this with an AtomicReference?

cpmeister replied:
Stick with the locks for now if you aren't comfortable using AtomicReferences. But yes, the way to do it would be with updateAndGet, and yes it can look a little awkward.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+    private final Object hubOutLock = new Object();
+    private final Object stateCommandsLock = new Object();
+    private final Object macAddressCommandLock = new Object();
+    private final Object zoneCommandLock = new Object();
```
Same as above

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,595 @@
+
+    private volatile @Nullable AdorneHubChangeNotify changeListener;
+
+    private final Object stopLock = new Object();
```
Good point - removed stop and replaced it with isCancelled. The stopLock however I do need to ensure hubConnector doesn't get swapped out while I am cancelling it (previously I had introduced it only as a perfromance optimization where a hubConnector swap would have slowed down repsonse to a stop request; now that I only rely on isCancelled it is not just an optimization but functionally required). 
Side note: there was a hole with cancelling commands upon stop - fixed that.

theiding commented:
@cpmeister , made most the requested changes. A couple of the requests are still unresolved though - I left my feedback as to why. Please take a look. Thanks!

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -82,56 +84,83 @@
-                           // troubleshooting
-
-    private volatile @Nullable AdorneHubChangeNotify changeListener;
+    private AtomicInteger commandId; // We assign increasing command ids to all REST commands to the hub for easier
```
Should have known better by now - sorry. Done. Made hubHost and hubControllerConnected final as well.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHubHandler.java:
```diff
@@ -0,0 +1,135 @@
+ */
+@NonNullByDefault
+public class AdorneHubHandler extends BaseBridgeHandler implements AdorneHubChangeNotify {
+    private Logger logger = LoggerFactory.getLogger(AdorneHubHandler.class);
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneSwitchHandler.java:
```diff
@@ -0,0 +1,172 @@
+        if (configZoneId != null) {
+            zoneId = configZoneId;
+        } else {
+            throw new IllegalStateException("zoneId must not be null as it is a required configuration parameter");
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/handler/AdorneHandlerFactory.java:
```diff
@@ -0,0 +1,74 @@
+@NonNullByDefault
+@Component(configurationPid = "binding.adorne", service = ThingHandlerFactory.class)
+public class AdorneHandlerFactory extends BaseThingHandlerFactory {
+    private Logger logger = LoggerFactory.getLogger(AdorneHandlerFactory.class);
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,588 @@
+    private @Nullable Connection hubConnection;
+    private CompletableFuture<@Nullable Void> hubControllerConnected;
+    private int hubReconnectSleep; // Sleep time before we attempt re-connect
+    private ScheduledExecutorService scheduler;
```
Done

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,588 @@
+        }
+    }
+
+    private @Nullable JsonObject getHubMsg() {
```
I felt splitting the connectivity functionality further would just introduce too much code overhead, but it does lead to nice separation of responsibility. Now that the connection class is no longer just a pojo I put it in its own file.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,588 @@
+            if (commandId == Integer.MAX_VALUE) {
+                newCommandId = 0;
+            }
+            return (++newCommandId);
```
I never noticed. Cleaned it up across.

theiding commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,588 @@
+        return (hubController == null || hubController.isCancelled());
+    }
+
+    private boolean connect() {
```
I use the lock to ensure a clean closing and setting to null to prevent a subsequent write or cancel. Creating a new connection on the other hand is ok as there won't be a competing write or cancel when there is no connection.

theiding commented:
@cpmeister , changes pushed.

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -362,15 +333,19 @@ private void msgLoop() {
-        } catch (RuntimeException e) {
+        } catch (
+
+        RuntimeException e) {
```
format

theiding commented:
@cpmeister , fixed the code format.

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java:
```diff
@@ -0,0 +1,524 @@
+    private int hubReconnectSleep; // Sleep time before we attempt re-connect
+    private final ScheduledExecutorService scheduler;
+
+    private volatile boolean stopWhenCommandsServed; // Stop the controller once all pending commands have been served
```
Should this also prevent new commands from being accepted?

theiding replied:
A case could be made either way. I opted for the simpler path. My only use case is during discovery where I only have a fixed set of commands to send and don't have to worry about subsequent commands.

cpmeister commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/discovery/AdorneDiscoveryService.java:
```diff
@@ -0,0 +1,115 @@
+        AdorneHubController adorneHubController = new AdorneHubController(config, scheduler, this);
+
+        // Limit life of hub controller to the duration of discovery
+        adorneHubController.stopBy(System.currentTimeMillis() + DISCOVERY_TIMEOUT_SECONDS * 1000);
```
stopScan is automatically called when `DISCOVERY_TIMEOUT_SECONDS` elapses. So why not utilize that instead of using a stopBy time?

theiding replied:
I didn't know stopScan existed. Much simpler to leverage that. Done.
I'll propose an update to the binding documentation separately.

cpmeister commented:
@theiding I haven't experienced the formatting issue personally, but I have seen it happen to others. I don't know if there is a way to fix it.

theiding commented:
Thank you @cpmeister.
@J-N-K , anything left on your side?

theiding commented:
When looking into updating stopScan documentation I realized I should call super.stopScan - fixed that.

J-N-K commented:
I‘ll have a look this evening. Thanks for your cooperation and patience.

J-N-K commented on bundles/org.openhab.binding.adorne/README.md:
```diff
@@ -0,0 +1,113 @@
+
+## Supported Things
+
+| Thing Type|Description|
```
Please format the tables with https://www.tablesgenerator.com/markdown_tables

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubChangeNotify.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * The {@link AdorneHubChangeNotify} interface is used by the {@link AdorneHubController} to notify listeners
+ * about
+ * Adorne device status and hub connection changes.
```
check formatting

theiding replied:
Done

J-N-K commented on bundles/org.openhab.binding.adorne/src/main/resources/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,26 @@
+				<label>Host</label>
+				<description>
+					Host name or IP address.
+				</description>
```
```suggestion
				</description>
				<context>network_address</context>
```

theiding replied:
Done

theiding commented:
@J-N-K - thanks for the quick turnaround. Made the changes.

theiding commented:
@J-N-K , @cpmeister , this was my first PR and it was a great learning experience. Thank you very much for investing so much of your time and being so diligent with the review - I didn't expect anybody to spend more than a couple minutes reviewing addons. In my next submission I can hopefully apply all my lessons learned and make it easier for you guys!
